File: core/src/main/java/com/triangleleft/flashcards/ui/main/MainPresenter.java
package com.triangleleft.flashcards.ui.main;

import android.support.annotation.NonNull;

import com.annimon.stream.Optional;
import com.triangleleft.flashcards.di.scope.ActivityScope;
import com.triangleleft.flashcards.service.account.AccountModule;
import com.triangleleft.flashcards.service.settings.Language;
import com.triangleleft.flashcards.service.settings.UserData;
import com.triangleleft.flashcards.service.vocabular.VocabularyWord;
import com.triangleleft.flashcards.ui.common.presenter.AbstractPresenter;
import com.triangleleft.flashcards.ui.login.ViewState;
import com.triangleleft.flashcards.ui.vocabular.VocabularyNavigator;
import com.triangleleft.flashcards.util.FunctionsAreNonnullByDefault;
import com.triangleleft.flashcards.util.Utils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.Executor;

import javax.inject.Inject;
import javax.inject.Named;

@FunctionsAreNonnullByDefault
@ActivityScope
public class MainPresenter extends AbstractPresenter<IMainView, ViewState> implements VocabularyNavigator {

    private static final Logger logger = LoggerFactory.getLogger(MainPresenter.class);
    private final AccountModule accountModule;
    private Page currentPage;
    private String title;
    private Optional<VocabularyWord> currentWord = Optional.empty();
    private boolean isSettingsDialogShown;

    @Inject
    public MainPresenter(AccountModule accountModule, @Named(VIEW_EXECUTOR) Executor executor) {
        super(IMainView.class, executor);
        this.accountModule = accountModule;
    }

    @Override
    public void onCreate() {
        Optional<UserData> userData = accountModule.getUserData();
        Utils.checkState(userData.isPresent(), "Showing main screen without userdata present");
        // It's possible that we are using account that doesn't have any languages
        userData.get().getCurrentLearningLanguage().ifPresent(language -> title = language.getName());
        currentPage = Page.LIST;
    }

    @Override
    public void onBind(IMainView view) {
        super.onBind(view);
        switch (currentPage) {
            case LIST:
                view.showList();
                break;
            case WORD:
                view.showWord(currentWord);
                break;
            default:
                throw new IllegalStateException("Unknown state " + currentPage);
        }
        view.setTitle(title);
        if (isSettingsDialogShown) {
            view.showFlashcardsDialog();
        }
    }

    @Override
    public void showWord(@NonNull Optional<VocabularyWord> word) {
        logger.debug("showWord() called with: word = [{}]", word);
        currentPage = Page.WORD;
        currentWord = word;
        getView().showWord(word);
    }

    public void onBackPressed() {
        if (currentPage == Page.LIST) {
            // list is considered top-level screen, so pressing back here should close app
            getView().finish();
        } else {
            // Go to top-level screen
            currentPage = Page.LIST;
            getView().showList();
        }
    }

    public void onLanguageChanged(Language currentLanguage) {
        logger.debug("onLanguageChanged() called with: currentLanguage = [{}]", currentLanguage);
        title = currentLanguage.getName();
        getView().setTitle(title);
        getView().reloadList();
    }

    public void setSettingsDialogShown(boolean isShown) {
        isSettingsDialogShown = isShown;
    }

    private enum Page {
        LIST,
        WORD
    }
}

File: core/src/main/kotlin/com/triangleleft/flashcards/ui/main/MainPresenter.kt
package com.triangleleft.flashcards.ui.main

import com.triangleleft.flashcards.di.scope.ActivityScope
import com.triangleleft.flashcards.service.account.AccountModule
import com.triangleleft.flashcards.service.settings.Language
import com.triangleleft.flashcards.service.vocabular.VocabularyWord
import com.triangleleft.flashcards.ui.common.presenter.AbstractPresenter
import com.triangleleft.flashcards.ui.vocabular.VocabularyNavigator
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject
import org.slf4j.LoggerFactory
import javax.inject.Inject

@ActivityScope
class MainPresenter @Inject
constructor(accountModule: AccountModule)
    : AbstractPresenter<IMainView, MainViewState>(), VocabularyNavigator {


    private val wordClicks = PublishSubject.create<VocabularyWord>()
    private val backPresses = PublishSubject.create<Any>()
    private val viewStates = BehaviorSubject.create<MainViewState>()
    private var disposable = CompositeDisposable()
    private val initialState: MainViewState

    init {
        val userData = accountModule.userData.get()

        // It's possible that we are using account that doesn't have any languages
        val title = userData.currentLearningLanguage?.name ?: ""

        initialState = MainViewState(title, false, MainViewState.Page.List)
    }

    override fun onCreate(savedViewState: MainViewState?) {
        val initialState = savedViewState ?: initialState

        Observable.merge(
                backPresses.map { MainViewActions.backress() },
                wordClicks.map { MainViewActions.wordClick(it) })
                .scan(initialState) { state, action -> action.reduce(state) }
                .distinctUntilChanged()
                .subscribe { viewStates.onNext(it) }
    }

    override fun showWord(word: VocabularyWord) {
        wordClicks.onNext(word)
    }

    override fun onRebind(view: IMainView) {
        super.onRebind(view)

        disposable = CompositeDisposable()
        // Connect view observables to our publish subjects
        // Connect view to our behavior subject
        disposable.addAll(
                viewStates.subscribe { view.render(it) },
                view.backPresses().subscribe { backPresses.onNext(it) },
                view.wordClicks().subscribe { wordClicks.onNext(it) }
        )
    }

    override fun onUnbind() {
        super.onUnbind()
        disposable.dispose()
    }

    fun onLanguageChanged(currentLanguage: Language) {
        logger.debug("onLanguageChanged() called with: currentLanguage = [{}]", currentLanguage)
//        title = currentLanguage.name
//        view.setTitle(title)
//        view.reloadList()
    }


    private enum class Page {
        LIST,
        WORD
    }

    companion object {

        private val logger = LoggerFactory.getLogger(MainPresenter::class.java)
    }
}
